File: src/main/java/io/github/saneea/feature/text/ToCharArray.java
package io.github.saneea.feature.text;

import java.io.IOException;
import java.io.Reader;
import java.io.Writer;

import io.github.saneea.Feature;
import io.github.saneea.FeatureContext;

public class ToCharArray implements//
		Feature, //
		Feature.In.Text.Reader, //
		Feature.Out.Text.Writer {

	private Reader in;
	private Writer out;

	@Override
	public Meta meta(FeatureContext context) {
		return Meta.from("convert input string to char array");
	}

	@Override
	public void run(FeatureContext context) throws IOException {
		int charCode;
		boolean first = true;

		while ((charCode = in.read()) != -1) {

			if (first) {
				first = false;
			} else {
				out.append(", ");
			}

			out.append(//
					wrapInQuotes(//
							wrapIfEscaped(//
									(char) charCode)));

		}
	}

	private String wrapInQuotes(String s) {
		return "'" + s + "'";
	}

	private String wrapIfEscaped(char c) {
		switch (c) {
		case '\0':
			return "\\0";
		case '\r':
			return "\\r";
		case '\n':
			return "\\n";
		case '\t':
			return "\\t";
		case '\f':
			return "\\f";
		case '\b':
			return "\\b";

		case '\'':
		case '\\':
			return "\\" + c;

		default:
			return "" + c;
		}
	}

	@Override
	public void setIn(Reader in) {
		this.in = in;
	}

	@Override
	public void setOut(Writer out) {
		this.out = out;
	}

}

File: src/main/java/io/github/saneea/dvh/feature/text/ToCharArray.kt
package io.github.saneea.dvh.feature.text

import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.FeatureContext
import java.io.Reader
import java.io.Writer

class ToCharArray :
    Feature,
    Feature.In.Text.Reader,
    Feature.Out.Text.Writer {

    override lateinit var context: FeatureContext
    override lateinit var inTextReader: Reader
    override lateinit var outTextWriter: Writer

    override val meta = Meta("convert input string to char array")

    override fun run() {
        var charCode: Int
        var first = true
        while (inTextReader.read().also { charCode = it } != -1) {
            if (first) {
                first = false
            } else {
                outTextWriter.append(", ")
            }
            outTextWriter.append(
                wrapInQuotes(
                    wrapIfEscaped(
                        charCode.toChar()
                    )
                )
            )
        }
    }

    private fun wrapInQuotes(s: String): String {
        return "'$s'"
    }

    private fun wrapIfEscaped(c: Char) =
        when (c) {
            '\u0000' -> "\\0"
            '\r' -> "\\r"
            '\n' -> "\\n"
            '\t' -> "\\t"
            '\b' -> "\\b"
            '\'', '\\' -> "\\" + c
            else -> "" + c
        }
}